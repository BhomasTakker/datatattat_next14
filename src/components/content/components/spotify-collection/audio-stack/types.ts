import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
