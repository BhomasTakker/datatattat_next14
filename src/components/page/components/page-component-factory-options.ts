export enum PageComponentsOptions {
	STACK = "Stack",
	GRID = "Grid",
}
