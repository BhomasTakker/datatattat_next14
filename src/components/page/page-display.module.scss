@use "../../scss/abstracts/units/" as *;

.page {
	display: flex;
	flex-direction: column;
}
