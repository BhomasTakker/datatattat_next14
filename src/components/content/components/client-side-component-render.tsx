"use client";

import { PageComponent } from "@/types/page";
import { UnknownObject } from "@/types/utils";
import { useEffect, useState } from "react";
import { ComponentType } from "./component-map";
import { With } from "@/types/component";

type ClientSideComponentProps = {
	Component: ComponentType;
	component: PageComponent;
	getData: (queryObject: With) => Promise<UnknownObject>;
};

export const ClientSideComponent = ({
	getData,
	Component,
	component,
}: ClientSideComponentProps) => {
	const { _with: componentQueryObject, componentProps } = component;
	const { _with: componentPropsQueryObject } = componentProps || {};
	const [componentData, setComponentData] = useState<UnknownObject | null>(
		null
	);
	const [isError, setIsError] = useState<boolean>(false);

	useEffect(() => {
		const getComponentData = async () => {
			try {
				const data = await getData(
					componentQueryObject || componentPropsQueryObject
				);
				setComponentData(data);
			} catch {
				setIsError(true);
			}
		};
		getComponentData();
	}, []);

	if (isError) {
		return <div>Error loading data</div>;
	}

	if (!componentData) {
		return <div>Loading...</div>;
	}

	return (
		<Component component={component} dataObject={{ data: componentData }} />
	);
};
