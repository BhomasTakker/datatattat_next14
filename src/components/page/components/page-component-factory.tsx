import { PageContent } from "@/types/page";
import {
	PageContainersMap,
	PageComponentsOptions,
} from "./page-component-factory-map";

export const PageComponentFactory = ({ content }: { content: PageContent }) => {
	const { container } = content || {};
	const { containerType } = container;

	const ContainerComponent = PageContainersMap.get(
		containerType as PageComponentsOptions
	);

	// do better error handling
	if (!ContainerComponent) {
		return <div>Component not found</div>;
	}

	return <ContainerComponent content={content} />;
};
