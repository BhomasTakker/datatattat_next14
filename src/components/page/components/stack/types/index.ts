import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
