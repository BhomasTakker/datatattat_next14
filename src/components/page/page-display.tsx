import { IPage } from "@/types/page";
import { PageProfile } from "./profile/page-profile";
import { PageComponentFactory } from "./components/page-component-factory";
import styles from "./page-display.module.scss";
import { getCurrentRoute } from "@/utils/route";

export const PageDisplay = async ({ page }: { page: IPage }) => {
	const { content, meta, profile } = page || {};

	// not here
	const currentRoute = await getCurrentRoute();
	const isUserRoute = currentRoute?.includes("/users/") || false;
	const isClient = isUserRoute;

	return (
		<main className={styles.page}>
			{/* REPLACE with meta function */}
			{/* Page head for meta data, seo, etc */}
			{/* <PageHead headData={meta} /> */}
			{/* Page profile data for title, description, createdBy, etc */}
			<PageProfile profile={profile} />
			<PageComponentFactory content={content} isClient={isClient} />
		</main>
	);
};
