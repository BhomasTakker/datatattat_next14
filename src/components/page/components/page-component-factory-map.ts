import { PageGrid } from "./grid/page-grid";
import { PageComponentsOptions } from "./page-component-factory-options";
import { PageStack } from "./stack/page-stack";

type PageComponents = typeof PageStack | typeof PageGrid;

// Where should this go? / here actually causes issues

export const PageContainersMap = new Map<PageComponentsOptions, PageComponents>(
	[
		[PageComponentsOptions.STACK, PageStack],
		[PageComponentsOptions.GRID, PageGrid],
	]
);
