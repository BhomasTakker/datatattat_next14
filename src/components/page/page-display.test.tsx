import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";

import { PageDisplay } from "./page-display";
import { PageContent, PageProfile } from "@/types/page";
import { PageComponentsOptions } from "./components/page-component-factory-options";

type Meta = {
	pageTitle: string;
	description: string;
};

jest.mock("./head/page-head", () => {
	return {
		__esModule: true,
		PageHead: ({
			headData,
			children,
		}: {
			headData: Meta;
			children: Array<React.ReactElement<any>>;
		}) => {
			const pageTitle = headData.pageTitle;
			document.title = pageTitle;
			return <div data-testid="test-page-head">{children}</div>;
		},
	};
});

jest.mock("./profile/page-profile", () => {
	return {
		__esModule: true,
		PageProfile: ({ profile }: { profile: PageProfile }) => {
			return <div data-testid="test-page-profile">{profile.pageTitle}</div>;
		},
	};
});

jest.mock("./components/page-component-factory", () => {
	return {
		__esModule: true,
		PageComponentFactory: ({ content }: { content: PageContent }) => {
			const { containerType } = content || {};

			return (
				<div data-testid="test-page-component-factory">{containerType}</div>
			);
		},
	};
});

const MOCK = {
	meta: {
		pageTitle: "My Page Title",
	},
	style: {},
	profile: {
		pageTitle: "My Page Title",
		pageTitleVariant: "variant",
		showPageTitle: true,
	},
	route: "",
	creator: {},
	content: {
		containerType: PageComponentsOptions.STACK,
		props: {},
		components: [],
	},
	createdAt: new Date(),
	updatedAt: new Date(),
};

describe("PageDisplay", () => {
	it("renders main elelment", () => {
		render(<PageDisplay page={MOCK} />);
		const main = screen.getByRole("main");
		expect(main).toBeInTheDocument();
	});

	it.skip("renders correct document title", () => {
		render(<PageDisplay page={MOCK} />);
		expect(document.title).toEqual("My Page Title");
	});

	it.skip("renders page head", () => {
		render(<PageDisplay page={MOCK} />);
		const pageHead = screen.getByTestId("test-page-head");
		expect(pageHead).toBeInTheDocument();
	});

	it("renders page profile", () => {
		render(<PageDisplay page={MOCK} />);
		const pageProfile = screen.getByTestId("test-page-profile");
		expect(pageProfile).toBeInTheDocument();
	});

	it("renders page component factory with content", () => {
		render(<PageDisplay page={MOCK} />);
		const pageComponentFactory = screen.getByTestId(
			"test-page-component-factory"
		);
		expect(pageComponentFactory).toBeInTheDocument();
	});

	it("renders PageDisplay unchanged", () => {
		const { container } = render(<PageDisplay page={MOCK} />);
		expect(container).toMatchSnapshot();
	});
});
