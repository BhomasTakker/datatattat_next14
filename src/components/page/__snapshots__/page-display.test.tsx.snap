// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PageDisplay renders PageDisplay unchanged 1`] = `
<div>
  <main
    class="page"
  >
    <div
      data-testid="test-page-profile"
    >
      My Page Title
    </div>
    <div
      data-testid="test-page-component-factory"
    >
      Stack
    </div>
  </main>
</div>
`;
